Document `program-options` (#10173)
haskell/cabal: doc/config.rst

+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``ghc-pkg``           |                                                                                                                                    |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``ghcjs``             |                                                                                                                                    |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``ghcjs-pkg``         |                                                                                                                                    |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | ``greencard``         | `<https://hackage.haskell.org/package/greencard>`_                                                                                 |

Reviewer: `greencard` doesn't appear to have been updated since 2014.

Author: You mean I should probably add some comment there saying that it is probably an outdated tool, right?

Reviewer: Yes, as with `haskell-suite`.

+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``haddock``           | `<https://haskell-haddock.readthedocs.io/latest/>`_                                                                                |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``happy``             | `<https://haskell-happy.readthedocs.io/en/latest/>`_                                                                               |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | ``haskell-suite``     | Haskell suite was abandoned long time ago.                                                                                         |

Reviewer: ```suggestion
| ``haskell-suite``     | Haskell suite was abandoned a long time ago.                                                                                         |
```
And similarly for the next several lines.

+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``hsc2hs``            | `<https://hackage.haskell.org/package/hsc2hs>`_                                                                                    |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``hscolour``          | `<https://hackage.haskell.org/package/hscolour>`_                                                                                  |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``jhc``               | `<http://repetae.net/computer/jhc/>`_                                                                                              |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | ``ld``                | Usually provided by GHC's ``"ld command"`` entry in ``ghc --info``.                                                                |

Reviewer: Is it worth noting here that `ld` is only used by `ghci` and Template Haskell, and to build shared objects? Most things link via the C compiler, so that they get the C runtime and any global constructors/destructors required by that.

Author: I think I kind of mention that in the warning below. I.e that these programs are sometimes invoked via GHC and not directly so these options might not apply.

With the specific case of ld-options being applied also when calling GHC as `-optl`.

+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+
| ``uhc``               | `<https://github.com/UU-ComputerScience/uhc>`_                                                                                     |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------------+

.. warning::

  It is important to not confuse these options with the ones listed in the
  :ref:`build info<build-info>` section. The ``*-options`` fields mentioned are in
  fact syntactic sugar for specific ``ghc-options`` that will be passed only on
  certain phases.

.. warning::

  These options will be used when ``cabal`` invokes the tool as part of the build process or as part of a
  :pkg-field:`build-tool-depends` declaration, not whenever the tool is invoked by
  third parties.

  In particular this means that for example ``gcc-options`` will be used when ``cabal``

>   invokes ``gcc``, which is **not** when C sources are compiled by GHC (eventhough GHC

Reviewer: ```suggestion
  invokes ``gcc``, which is **not** when C sources are compiled by GHC (even though GHC
```

  might invoke ``gcc`` internally). In order to provide options through GHC for those programs, one has to check the
  GHC User guide's `Section <https://downloads.haskell.org/ghc/latest/docs/users_guide/phases.html#forcing-options-to-a-particular-phase>`_.
  In short, those options have to be given as ``-opt<phase>`` flags to GHC.

.. note::

  The only case that violates the rule specified in this last warning above is
  ``ld-options``, which get passed as ``-optl`` options when GHC is invoked for

>   linking, same as the :pkg-field:`ld-options` field in package descriptions.

Reviewer: ```suggestion
  linking, as with the :pkg-field:`ld-options` field in package descriptions.
```
